gtsam/hybrid/DCFactor.h

```diff
@@ -0,0 +1,264 @@
+   */
+  virtual double error(const gtsam::Values& continuousVals,
+                       const gtsam::DiscreteValues& discreteVals) const = 0;
+
+  /**
+   * Linearize the error function with respect to the continuous
+   * variables (given in `keys_`) at the point specified by `continuousVals`.
+   * Since this Jacobian can be dependent on the assignment to discrete
+   * variables as well, they are required as a parameter.
+   *
+   * This will be specific to the model being implemented, so it is a pure
+   * virtual function and must be overridden.
+   *
+   * @param continuousVals - Assignment to the continuous variables in `keys_`
+   * @param discreteVals - Likewise, assignment to the discrete variables in
+   * `discreteKeys__`.
+   */
+  virtual boost::shared_ptr<gtsam::GaussianFactor> linearize(
+      const gtsam::Values& continuousVals,
+      const DiscreteValues& discreteVals) const = 0;
+
+  /**
+   * Returns true when the DCFactor is equal to `other`
+   *
+   * This will be specific to the model being implemented, so it is a pure
+   * virtual function and must be overridden.
+   *
+   * @param other - the DCFactor for which to test equality.
+   * @param tol = 1e-9 - numerical tolerance for any floating point comparisons.
+   */
+  virtual bool equals(const DCFactor& other, double tol = 1e-9) const = 0;
+
+  /**
+   * Returns the number of rows in the Jacobian with respect to the continuous
+   * variables for this factor. Internally this is used in the conversion to a
+   * gtsam::NonlinearFactor
+   *
+   * TODO(kevin): not sure if needed???
+   *
+   * This will be specific to the model being implemented, so it is a pure
+   * virtual function and must be overridden
+   *
+   * @return the dimension (number of rows) in the Jacobian of the `error`
+   * function with respect to continuous variables for
+   */
+  virtual size_t dim() const = 0;
+
+  /*
+   * Return the discrete keys for this factor.
+   */
+  gtsam::DiscreteKeys discreteKeys() const { return discreteKeys_; }
+
+  /**
+   * Converts the DCFactor to a gtsam::DecisionTreeFactor. Internally, this will
+   * be used to generate a gtsam::DiscreteFactor type, which itself requires a
+   * conversion function to gtsam::DecisionTreeFactor for inference using GTSAM.
+   *
+   * Performing this conversion can be problem specific, so we allow for the
+   * option to override, but try to implement a sensible default: we assume the
+   * error function can be called setting each discrete key's variable
+   * individually, and the overall DCFactor can itself be factored as a product
+   * of unary gtsam::DecisionTreeFactors.
+   *
+   * Alternative implementations might consider something like the
+   * gtsam::AllDiff approach here:
+   * https://github.com/borglab/gtsam/blob/43e8f1e5aeaf11890262722c1e5e04a11dbf9d75/gtsam_unstable/discrete/AllDiff.cpp#L43
+   *
+   * @param continuousVals - an assignment to the continuous variables
+   * @param discreteVals -
```

Comment:
What is the role for discreteVals here? @keevindoherty  should provide this knowledge :-)

Reply:
`discreteVals` is an assignment to the discrete variables. This is useful for a kind of discrete-continuous smart factor where we've eliminated some subset of variables conditioned on a joint discrete and continuous assignment. In our case, we are using these for semantic+geometric measurements with ambiguous associations in semantic SLAM. For max-product elimination, check out [this max-mixture factor](https://github.com/MarineRoboticsGroup/dcsam/blob/main/include/dcsam/DCMaxMixtureFactor.h) for example, or for sum-product, we have [a discrete-continuous EMFactor here](https://github.com/MarineRoboticsGroup/dcsam/blob/main/include/dcsam/DCEMFactor.h). There are perhaps other applications I haven't thought about yet. Not all DCFactors have to use these values, but we pass them in case they are needed.